File: python/packages/ob-util/src/ob_util/export.py
"""DEP-5 copyright export for OriginBlame data."""

from __future__ import annotations

from pathlib import Path

from ob.authors import get_author
from ob.exceptions import OBStorageError
from ob.indexer import list_pid_files, read_all_manifest
from ob.register import get_section
from ob.util import find_ob_dir


def export_copyright(
    output: str | None = None,
    data_files: list[str] | None = None,
    ob_dir: Path | None = None,
) -> str:
    """Export copyright information in DEP-5 format.

    Reads manifest records, resolves section/author metadata, and produces
    DEP-5 copyright blocks.  Does NOT log export operations.

    Args:
        output: Output file path.  If None, returns the string.
        data_files: Filter by data file paths.  If None, exports all.
        ob_dir: Repository root directory.  If None, auto-detected.

    Returns:
        DEP-5 formatted copyright string.

    Raises:
        OBStorageError: If unmerged PID files are detected.
    """
    if ob_dir is None:
        ob_dir = find_ob_dir()

    pid_files = list_pid_files(ob_dir)
    if pid_files:
        raise OBStorageError(
            "Unmerged PID files detected. Run 'ob clean' first."
        )

    records = read_all_manifest(ob_dir)

    if data_files:
        data_files_set = set(data_files)
        records = [r for r in records if r.get("file") in data_files_set]

    blocks: list[str] = []
    for record in records:
        line_hash = record.get("line_hash", "")
        file_path = record.get("file", "")
        sources = record.get("sources", [])

        if not line_hash or not file_path:
            continue

        files_field = f"{file_path}:{line_hash}"

        for section_hash in sources:
            section = get_section(ob_dir, section_hash)
            if section is None:
                continue

            license_name = section.get("license") or "UNKNOWN"
            year = section.get("year", "")
            author_ids = section.get("authors", [])

            copyright_lines: list[str] = []
            for author_id in author_ids:
                author = get_author(ob_dir, author_id)
                if author is not None:
                    name = author.get("name", "")
                    email = author.get("email", "")
                    if year:
                        copyright_lines.append(f"{year} {name} <{email}>")
                    else:
                        copyright_lines.append(f"{name} <{email}>")

            if not copyright_lines:
                continue

            copyright_field = "\n".join(
                f"Copyright: {line}" for line in copyright_lines
            )

            block = (
                f"Files: {files_field}\n"
                f"{copyright_field}\n"
                f"License: {license_name}\n"
            )
            blocks.append(block)

    result = "\n".join(blocks)

    if output:
        output_path = Path(output)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        output_path.write_text(result, encoding="utf-8")

    return result

File: src/export.rs
use anyhow::Result;
use std::path::Path;

pub fn export_copyright(
    ob_dir: &Path,
    output: Option<&Path>,
    data_files: Option<&[String]>,
) -> Result<String> {
    let pid_files = crate::indexer::list_pid_files(ob_dir)?;
    anyhow::ensure!(
        pid_files.is_empty(),
        "Unmerged PID files found. Run `ob clean` first."
    );

    let manifests = crate::indexer::read_all_documents(ob_dir)?;
    let sections = crate::storage::shard_iterate_all(ob_dir, "sections")?;
    let authors = crate::storage::shard_iterate_all(ob_dir, "authors")?;

    let mut section_map = std::collections::HashMap::new();
    for s in &sections {
        if let Some(hash) = s.get("section_hash").and_then(|v| v.as_str()) {
            section_map.insert(hash.to_string(), s.clone());
        }
    }

    let mut author_map = std::collections::HashMap::new();
    for a in &authors {
        if let Some(id) = a.get("id").and_then(|v| v.as_str()) {
            author_map.insert(id.to_string(), a.clone());
        }
    }

    let mut blocks = Vec::new();
    for m in &manifests {
        let file_val = m.get("file").and_then(|v| v.as_str()).unwrap_or("");
        let line_hash = m.get("line_hash").and_then(|v| v.as_str()).unwrap_or("");

        if let Some(filter) = data_files {
            if !filter.iter().any(|f| f == file_val) {
                continue;
            }
        }

        let revoked = m.get("revoked").and_then(|v| v.as_bool()).unwrap_or(false);
        if revoked {
            continue;
        }

        let sources = match m.get("sources").and_then(|v| v.as_array()) {
            Some(s) => s,
            None => continue,
        };

        for src in sources {
            let section_hash = match src.as_str() {
                Some(s) => s,
                None => continue,
            };
            let section = match section_map.get(section_hash) {
                Some(s) => s,
                None => continue,
            };
            let section_revoked = section
                .get("revoked")
                .and_then(|v| v.as_bool())
                .unwrap_or(false);
            if section_revoked {
                continue;
            }

            let license = section
                .get("license")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            let year = section.get("year").and_then(|v| v.as_str()).unwrap_or("");

            if let Some(section_authors) = section.get("authors").and_then(|v| v.as_array()) {
                for aid in section_authors {
                    let author_id = match aid.as_str() {
                        Some(s) => s,
                        None => continue,
                    };
                    let author = match author_map.get(author_id) {
                        Some(a) => a,
                        None => continue,
                    };
                    let author_revoked = author
                        .get("revoked")
                        .and_then(|v| v.as_bool())
                        .unwrap_or(false);
                    if author_revoked {
                        continue;
                    }
                    let name = author
                        .get("name")
                        .and_then(|v| v.as_str())
                        .unwrap_or("Unknown");
                    let email = author.get("email").and_then(|v| v.as_str()).unwrap_or("");

                    let mut block = String::new();
                    block.push_str(&format!("Files: {}:{}\n", file_val, line_hash));
                    if !year.is_empty() {
                        block.push_str(&format!("Copyright: {} {} <{}>\n", year, name, email));
                    } else {
                        block.push_str(&format!("Copyright: {} <{}>\n", name, email));
                    }
                    block.push_str(&format!("License: {}\n", license));
                    blocks.push(block);
                }
            }
        }
    }

    let result = blocks.join("\n");
    if let Some(output_path) = output {
        std::fs::write(output_path, &result)?;
    }
    Ok(result)
}
